src/Cli/dotnet/OptionForwardingExtensions.cs

```diff
@@ -18,10 +19,45 @@ public static class OptionForwardingExtensions
 
         public static ForwardedOption<T> ForwardAsSingle<T>(this ForwardedOption<T> option, Func<T, string> format) => option.SetForwardingFunction(format);
 
+        private static Parser<(string key, string value)[]> KeyValueParser {
+            get {
+                bool IsValidMSBuildPropertyChar (char c) => Char.IsLetterOrDigit(c) || (c == '_') || (c == '-');
+            var keyParser = 
+                Parse.Char(IsValidMSBuildPropertyChar, "Any alphanumeric character or '_'").AtLeastOnce().Text().Named("key");
+
+                var unquotedValueParser = Parse.CharExcept('"').Named("unquoted value");
+            
+                var quoteParser = Parse.Char('"').Named("quote");
+                var semiParser = Parse.Char(';').Named("tick");
+                var quotedValueParser = 
+                    Parse.Contained(unquotedValueParser.AtLeastOnce(), quoteParser, quoteParser).Text().Select(s => $"\"{s}\"").Named("quoted value");
+                var valueParser = 
+                    Parse.XOr(quotedValueParser, Parse.CharExcept(';').AtLeastOnce().Text()).Named("value");
+                var keyValueParser = 
+                    from key in keyParser
+                    from _equals in Parse.Char('=')
+                    from value in valueParser
+                    select (key, value);
+                var semiOrCommaParser = Parse.Chars(new[]{';', ','});
+                var keyValues = Parse.DelimitedBy(keyValueParser.Named("key/value"), Parse.Char(';')).Select(pairs => pairs.ToArray()).Named("key/value list");
+                return keyValues;
```

Comment:
What are the performance characteristics of this? This feels like a very expensive way to create a simple parser.

Reply:
I was very hesitant to roll my own parser - I'm not an expert at doing that, and using a library to handle the combinations adds nice guardrails.

However I've rolled a new one that I'll need to test the heck out of now to sidestep the issue.